src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs

```diff
@@ -335,64 +360,102 @@ private SuggestedActionSet OrganizeRefactorings(Workspace workspace, CodeRefacto
+                            cancellationToken).ConfigureAwait(false);
+
+                        if (result.HasFix)
+                        {
+                            Logger.Log(FunctionId.SuggestedActions_HasSuggestedActionsAsync);
+                            return true;
+                        }
+
+                        if (result.PartialResult)
+                        {
+                            // reset solution version number so that we can raise suggested action changed event
+                            Volatile.Write(ref _lastSolutionVersionReported, InvalidSolutionVersion);
+                            return false;
+                        }
+                    }
                 }
 
-                var document = documentAndSnapshot.Value.Item1;
+                return false;
+            }
 
-                // make sure current document support codefix
-                var supportCodeFix = document.Project.Solution.Workspace.Services.GetService<IDocumentSupportsSuggestionService>();
-                if (!supportCodeFix.SupportsCodeFixes(document))
-                {
-                    return null;
+            private async Task<bool> HasRefactoringsAsync(
+                IOptionService optionService,
+                IDocumentSupportsSuggestionService supportSuggestion,
+                ISuggestedActionCategorySet requestedActionCategories,
+                Document document, ITextSnapshot snapshot, SnapshotSpan range,
+                CancellationToken cancellationToken)
+            {
+                if (optionService.GetOption(EditorComponentOnOffOptions.CodeRefactorings) &&
+                    _owner._codeRefactoringService != null &&
+                    supportSuggestion.SupportsCodeFixes(document) &&
+                    (requestedActionCategories.Contains(PredefinedSuggestedActionCategoryNames.Refactoring) ||
+                     requestedActionCategories.Contains(PredefinedSuggestedActionCategoryNames.Any)))
+                {
+                    var selection = TryGetCodeRefactoringSelection(snapshot, range);
+                    if (selection.HasValue)
+                    {
+                        return await Task.Run(
+                            async () => await _owner._codeRefactoringService.HasRefactoringsAsync(
+                                document, selection.Value, cancellationToken).ConfigureAwait(false),
+                            cancellationToken).ConfigureAwait(false);
+                    }
                 }
 
-                return await service.GetFirstDiagnosticWithFixAsync(document, range.Span.ToTextSpan(), cancellationToken).ConfigureAwait(false);
+                return false;
             }
 
-            private TextSpan? GetCodeRefactoringSelection(ITextSnapshot snapshot)
+            private TextSpan? TryGetCodeRefactoringSelection(ITextSnapshot snapshot, SnapshotSpan range)
             {
                 var selectedSpans = _textView.Selection.SelectedSpans
```

Comment:
Thanks @heejaechang. Looks like we are already ensuring that the snapshots are the same within ```GetMatchingDocumentAndSnapshotAsync()``` below. I'll change ```TryGetCodeRefactoringSelection()``` to only pass in the range argument